internal/pkg/service/common/etcdclient/etcdclient.go

```diff
@@ -0,0 +1,228 @@
+func WithConnectTimeout(v time.Duration) Option {
+	return func(c *config) {
+		c.connectTimeout = v
+	}
+}
+
+// WithDialTimeout defines the maximum time of one connection attempt.
+// In case of failure, a retry follow.
+func WithDialTimeout(v time.Duration) Option {
+	return func(c *config) {
+		c.connectTimeout = v
+	}
+}
+
+func WithKeepAliveTimeout(v time.Duration) Option {
+	return func(c *config) {
+		c.keepAliveTimeout = v
+	}
+}
+
+func WithKeepAliveInterval(v time.Duration) Option {
+	return func(c *config) {
+		c.keepAliveInterval = v
+	}
+}
+
+// WithAutoSyncInterval defines how often the list of cluster nodes/endpoints will be synced.
+// This is useful if the cluster will scale up or down.
+func WithAutoSyncInterval(v time.Duration) Option {
+	return func(c *config) {
+		c.keepAliveTimeout = v
+	}
+}
+
+func WithLogger(v log.Logger) Option {
+	return func(c *config) {
+		c.logger = v
+	}
+}
+
+// WithWaitGroup attach the client to a waitGroup.
+// When the connection is established, the waitGroup counter is increased,
+// when the connection is closed, it is decreased.
+// Connection is closed when the context is done, see New.
+func WithWaitGroup(v *sync.WaitGroup) Option {
+	return func(c *config) {
+		c.waitGroup = v
+	}
+}
+
+// New creates new etcd client.
+// The client terminates the connection when the context is done.
+func New(ctx context.Context, tracer trace.Tracer, endpoint, namespace string, opts ...Option) (c *etcd.Client, err error) {
+	ctx, span := tracer.Start(ctx, "kac.api.server.templates.dependencies.EtcdClient")
+	defer telemetry.EndSpan(span, &err)
+
+	// Apply options
+	conf := config{
+		endpoint:          endpoint,
+		namespace:         namespace,
+		connectCtx:        ctx,
+		connectTimeout:    defaultConnectionTimeout,
+		keepAliveTimeout:  defaultKeepAliveTimeout,
+		keepAliveInterval: defaultKeepAliveInterval,
+		autoSyncInterval:  defaultAutoSyncInterval,
+	}
+	for _, o := range opts {
+		o(&conf)
+	}
+
+	// Trim and validate
+	endpoint = strings.Trim(endpoint, " /")
+	if endpoint == "" {
+		return nil, errors.New("etcd endpoint is not set")
+	}
+	namespace = strings.Trim(namespace, " /") + "/"
+	if namespace == "/" {
+		return nil, errors.New("etcd namespace is not set")
+	}
+
+	// Setup logger
+	logger := conf.logger
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	// Create a zap logger for etcd client
+	encoder := zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig())
+	etcdLogger := zap.New(log.NewCallbackCore(func(entry zapcore.Entry, fields []zapcore.Field) {
+		// Add component=etcd-client field
+		fields = append(fields, zapcore.Field{Key: "component", String: "etcd-client", Type: zapcore.StringType})
+
+		// Encode and log message
+		if bytes, err := encoder.EncodeEntry(entry, fields); err == nil {
+			logger.Log(entry.Level.String(), strings.TrimRight(bytes.String(), "\n"))
+		} else {
+			logger.Warnf("cannot log msg from etcd client: %s", err)
+		}
+	}))
```

Comment:
Doteraz sme etcd klientovi nenastavovali logger, tak logoval na `stdout/stderr`.
Teraz ide vsetko cez nas logger, pomoze nam to, ked sa prejde z PT na DD.